lib/storage/storage.go

```diff
@@ -801,6 +833,9 @@ func (s *Storage) mustRotateIndexDB(currentTime time.Time) {
 	nextRotationTimestamp := currentTime.Unix() + s.retentionMsecs/1000
 	s.nextRotationTimestamp.Store(nextRotationTimestamp)
 
+	// Update indexDBs under write lock. See idbsLocked().
+	s.idbLock.Lock()
```

Comment:
For me, it looks like there is an issue with atomic reference count on indexdb.

 It's implemented incorrectly for current indexdb.

Since we don't incRef for current indexdb and indexdb rotation happens frequently. It's possible the case, when some API holds uses current indexdb. And due to rotation currIdx becomes prev and dropped at some point. But since we didn't incRef during current indexdb load, it drops without properly finished ingestion. 

 This PR fixes it only partially during metrics ingestion, but is it still possible the same situation during query?

Reply:
> For me, it looks like there is an issue with atomic reference count on indexdb.
> 
> It's implemented incorrectly for current indexdb.

Correct. The issues is that inc/dec ref is not done in all places where indexDB can be used. More precisely, in all possible execution paths (goroutines), such as main, signal hadlers, request handlers, and background workers. Currently the ref counter is incremented only in main (where the storage is opened), and decremented in background workers (indexdb rotation) and signal handler (stopping the storage). The request handlers do not inc/dec the counters and hence the problem with concurrent request handling and indexdb rotation.

> 
> Since we don't incRef for current indexdb and indexdb rotation happens frequently. It's possible the case, when some API holds uses current indexdb. And due to rotation currIdx becomes prev and dropped at some point. But since we didn't incRef during current indexdb load, it drops without properly finished ingestion.

Correct. This is exactly what this PR attempts to fix.

> This PR fixes it only partially during metrics ingestion, but is it still possible the same situation during query?

Correct. The PR only fixes that particular flaky test. And same thing with querying is possible, at least theoretically. I agree there should be a holistic solution but not sure if it should be implemented in context of this PR.

Reply:
I think it's too much changes for fixing flaky test. 

And it also looks like that test is written incorrectly - it assumes, that indexdb could be rotated frequently and tests impossible case.

 I propose to add the following changes into the TestStorageRorateIndexdb 
- add 3 rotation cycles
- rotate only once per ingestion cycle

Reply:
Done. Reverted the code in storage.go and changed the test per your recommendations. Thanks!